// class-validator-jsonschema/build/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.validationMetadatasToSchemas = exports.JSONSchema = void 0;
const cv = require("class-validator");
const _ = require("lodash");
const decorators_1 = require("./decorators");
const defaultConverters_1 = require("./defaultConverters");
const options_1 = require("./options");
var decorators_2 = require("./decorators");
Object.defineProperty(exports, "JSONSchema", { enumerable: true, get: function () { return decorators_2.JSONSchema; } });
function validationMetadatasToSchemas(userOptions) {
    const options = Object.assign(Object.assign({}, options_1.defaultOptions), userOptions);
    const metadatas = getMetadatasFromStorage(options.classValidatorMetadataStorage);
    const schemas = _(metadatas)
        .groupBy((value) => _.get(value, `target.${options.schemaNameField}`, _.get(value, 'target.name')))
        .mapValues((ownMetas) => {
        const target = ownMetas[0].target;
        const metas = ownMetas
            .concat(getInheritedMetadatas(target, metadatas))
            .filter((propMeta) => !isExcluded(propMeta, options));
        const properties = _(metas)
            .groupBy('propertyName')
            .mapValues((propMetas, propKey) => {
            const schema = applyConverters(propMetas, options);
            return applyDecorators(schema, target, options, propKey);
        })
            .value();
        const definitionSchema = {
            properties,
            type: 'object',
        };
        const required = getRequiredPropNames(target, metas, options);
        if (required.length > 0) {
            definitionSchema.required = required;
        }
        return applyDecorators(definitionSchema, target, options, target.name);
    })
        .value();
    return schemas;
}
exports.validationMetadatasToSchemas = validationMetadatasToSchemas;
function getMetadatasFromStorage(storage) {
    const metadatas = _.get(storage, 'validationMetadatas');
    const constraints = _.get(storage, 'constraintMetadatas');
    return metadatas.map((meta) => {
        if (meta.constraintCls) {
            const constraint = constraints.find((c) => c.target === meta.constraintCls);
            if (constraint) {
                return Object.assign(Object.assign({}, meta), { type: constraint.name });
            }
        }
        return meta;
    });
}
function getInheritedMetadatas(target, metadatas) {
    return metadatas.filter((d) => d.target instanceof Function &&
        target.prototype instanceof d.target &&
        !_.find(metadatas, {
            propertyName: d.propertyName,
            target,
            type: d.type,
        }));
}
function applyConverters(propertyMetadatas, options) {
    const converters = Object.assign(Object.assign({}, defaultConverters_1.defaultConverters), options.additionalConverters);
    const convert = (meta) => {
        var _a;
        const typeMeta = (_a = options.classTransformerMetadataStorage) === null || _a === void 0 ? void 0 : _a.findTypeMetadata(meta.target, meta.propertyName);
        const isMap = typeMeta &&
            typeMeta.reflectedType &&
            new typeMeta.reflectedType() instanceof Map;
        const converter = converters[meta.type] || converters[cv.ValidationTypes.CUSTOM_VALIDATION];
        const items = _.isFunction(converter) ? converter(meta, options) : converter;
        if (meta.each && isMap) {
            return {
                additionalProperties: Object.assign({}, items),
                type: 'object',
            };
        }
        return meta.each ? { items, type: 'array' } : items;
    };
    return _.merge({}, ...propertyMetadatas.map(convert));
}
function isExcluded(propertyMetadata, options) {
    var _a;
    return !!((_a = options.classTransformerMetadataStorage) === null || _a === void 0 ? void 0 : _a.findExcludeMetadata(propertyMetadata.target, propertyMetadata.propertyName));
}
function applyDecorators(schema, target, options, propertyName) {
    const additionalSchema = decorators_1.getMetadataSchema(target.prototype, propertyName);
    return _.isFunction(additionalSchema)
        ? additionalSchema(schema, options)
        : _.merge({}, schema, additionalSchema);
}
function getRequiredPropNames(target, metadatas, options) {
    function isDefined(metas) {
        return _.some(metas, { type: cv.ValidationTypes.IS_DEFINED });
    }
    function isOptional(metas) {
        return _.some(metas, ({ type }) => _.includes([cv.ValidationTypes.CONDITIONAL_VALIDATION, cv.IS_EMPTY], type));
    }
    return _(metadatas)
        .groupBy('propertyName')
        .pickBy((metas) => {
        const [own, inherited] = _.partition(metas, (d) => d.target === target);
        return options.skipMissingProperties
            ? isDefined(own) || (!isOptional(own) && isDefined(inherited))
            : !(isOptional(own) || isOptional(inherited));
    })
        .keys()
        .value();
}
